With Nx 15.2.x using Karma for @nrwl/angular is broken

### Current Behavior

While updating our Ionic Angular package, we found a problem with Karma (not sure how important that is these days).

This is the generated executor for Karma:
```
"test": {
  "executor": "@angular-devkit/build-angular:karma",
  "options": {
    "main": "apps/testkarma/src/test.ts",
    "tsConfig": "apps/testkarma/tsconfig.spec.json",
    "karmaConfig": "apps/testkarma/karma.conf.js",
    "polyfills": "apps/testkarma/src/polyfills.ts",
    "styles": [],
    "scripts": [],
    "assets": []
  }
}
```

and the tsconfig:
```
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "outDir": "../../dist/out-tsc",
    "types": ["jasmine", "node"]
  },
  "files": ["src/test.ts", "src/polyfills.ts"],
  "include": ["**/*.spec.ts", "**/*.test.ts", "**/*.d.ts"]
}
```

In both, there's a `polifills.ts` mentioned, which isn't there. I'm not sure if removing that line is enough here or if something else would be missing here.

### Expected Behavior

Karma is able to run.

### Github Repo

_No response_

### Steps to Reproduce

1. Create an Nx workspace
2. Generate an Angular application using Karma
   - nx g @nrwl/angular:app testkarma --unitTestRunner=karma
3. Run the tests
  - nx test testkarma


### Nx Report

```shell
npx nx report

 >  NX   Report complete - copy this into the issue template

   Node : 16.18.0
   OS   : darwin x64
   yarn : 1.22.19

   nx : 15.2.0
   @nrwl/angular : 15.2.0
   @nrwl/cypress : 15.2.0
   @nrwl/detox : Not Found
   @nrwl/devkit : 15.2.0
   @nrwl/esbuild : Not Found
   @nrwl/eslint-plugin-nx : 15.2.0
   @nrwl/expo : Not Found
   @nrwl/express : Not Found
   @nrwl/jest : 15.2.0
   @nrwl/js : 15.2.0
   @nrwl/linter : 15.2.0
   @nrwl/nest : Not Found
   @nrwl/next : Not Found
   @nrwl/node : 15.2.0
   @nrwl/nx-cloud : 15.0.2
   @nrwl/nx-plugin : 15.2.0
   @nrwl/react : 15.2.0
   @nrwl/react-native : Not Found
   @nrwl/rollup : 15.2.0
   @nrwl/schematics : Not Found
   @nrwl/storybook : 15.2.0
   @nrwl/web : 15.2.0
   @nrwl/webpack : 15.2.0
   @nrwl/workspace : 15.2.0
   typescript : 4.8.4
   ---------------------------------------
   Local workspace plugins:
   ---------------------------------------
   Community plugins:
   	 @ionic/angular: 6.3.7
   	 @nxext/ionic-angular: 999.9.9
```


### Failure Logs

```shell
npx nx test testkarma

> nx run testkarma:test

⠙ Generating browser application bundles (phase: setup)...Error: ENOENT: no such file or directory, open '/Users/dominikpieper/Projects/angular/nxext/tmp/nx-e2e/proj/apps/testkarma/src/polyfills.ts'
    at Object.openSync (node:fs:590:3)
    at Module.readFileSync (node:fs:458:35)
    at NodeJSFileSystem.readFile (file:///Users/dominikpieper/Projects/angular/nxext/tmp/nx-e2e/proj/node_modules/@angular/compiler-cli/bundles/chunk-WEEZR5JR.js:333:16)
    at EsmDependencyHost.recursivelyCollectDependencies (file:///Users/dominikpieper/Projects/angular/nxext/tmp/nx-e2e/proj/node_modules/@angular/compiler-cli/bundles/chunk-3KIUHCHT.js:217:34)
    at EsmDependencyHost.processFile (file:///Users/dominikpieper/Projects/angular/nxext/tmp/nx-e2e/proj/node_modules/@angular/compiler-cli/bundles/chunk-3KIUHCHT.js:246:12)
    at EsmDependencyHost.collectDependenciesInFiles (file:///Users/dominikpieper/Projects/angular/nxext/tmp/nx-e2e/proj/node_modules/@angular/compiler-cli/bundles/chunk-3KIUHCHT.js:213:12)
    at ProgramBasedEntryPointFinder.getInitialEntryPointPaths (file:///Users/dominikpieper/Projects/angular/nxext/tmp/nx-e2e/proj/node_modules/@angular/compiler-cli/bundles/chunk-3KIUHCHT.js:864:10)
    at ProgramBasedEntryPointFinder.findEntryPoints (file:///Users/dominikpieper/Projects/angular/nxext/tmp/nx-e2e/proj/node_modules/@angular/compiler-cli/bundles/chunk-3KIUHCHT.js:823:35)
    at file:///Users/dominikpieper/Projects/angular/nxext/tmp/nx-e2e/proj/node_modules/@angular/compiler-cli/bundles/chunk-3KIUHCHT.js:1273:33
    at new ClusterMaster (file:///Users/dominikpieper/Projects/angular/nxext/tmp/nx-e2e/proj/node_modules/@angular/compiler-cli/bundles/chunk-3KIUHCHT.js:1375:22)
✔ Browser application bundle generation complete.

./apps/testkarma/src/app/app.component.spec.ts - Error: Module build failed (from ./node_modules/@ngtools/webpack/src/ivy/index.js):
Error: Emit attempted before Angular Webpack plugin initialization.

./apps/testkarma/src/test.ts - Error: Module build failed (from ./node_modules/@ngtools/webpack/src/ivy/index.js):
Error: Emit attempted before Angular Webpack plugin initialization.

Error: Module not found: Error: Can't resolve 'apps/testkarma/src/polyfills.ts' in '/Users/dominikpieper/Projects/angular/nxext/tmp/nx-e2e/proj'

Error: Failed to initialize Angular compilation - NGCC failed.


21 11 2022 16:41:44.158:INFO [karma-server]: Karma v6.4.1 server started at http://localhost:9876/
21 11 2022 16:41:44.160:INFO [launcher]: Launching browsers Chrome with concurrency unlimited
21 11 2022 16:41:44.160:ERROR [karma-server]: Error: Found 1 load error
    at Server.<anonymous> (/Users/dominikpieper/Projects/angular/nxext/tmp/nx-e2e/proj/node_modules/karma/lib/server.js:243:26)
    at Object.onceWrapper (node:events:627:28)
    at Server.emit (node:events:525:35)
    at Server.emit (node:domain:489:12)
    at emitListeningNT (node:net:1497:10)
    at processTicksAndRejections (node:internal/process/task_queues:82:21)
```


### Additional Information

_No response_

--- a/packages/angular/src/generators/karma-project/karma-project.spec.ts
+++ b/packages/angular/src/generators/karma-project/karma-project.spec.ts
@@ -1,5 +1,9 @@
 import type { Tree } from '@nrwl/devkit';
 import * as devkit from '@nrwl/devkit';
+import {
+  readProjectConfiguration,
+  updateProjectConfiguration,
+} from '@nrwl/devkit';
 import { createTreeWithEmptyV1Workspace } from '@nrwl/devkit/testing';
 import { karmaProjectGenerator } from './karma-project';
 import libraryGenerator from '../library/library';
@@ -46,8 +50,36 @@ describe('karmaProject', () => {
 
     expect(tree.exists('/libs/lib1/karma.conf.js')).toBeTruthy();
     expect(tree.exists('/libs/lib1/tsconfig.spec.json')).toBeTruthy();
+    expect(
+      tree.read('/libs/lib1/tsconfig.spec.json', 'utf-8')
+    ).toMatchSnapshot();
     expect(tree.exists('/libs/lib1/src/test.ts')).toBeTruthy();
     expect(tree.exists('karma.conf.js')).toBeTruthy();
+  });
+
+  it('should generate files and not add polyfills if it is using ng v15 style polyfills', async () => {
+    expect(tree.exists('karma.conf.js')).toBeFalsy();
+    await karmaProjectGenerator(tree, { project: 'app1' });
+
+    expect(tree.exists('/apps/app1/tsconfig.spec.json')).toBeTruthy();
+    expect(
+      tree.read('/apps/app1/tsconfig.spec.json', 'utf-8')
+    ).toMatchSnapshot();
+  });
+
+  it('should generate files and correctly add polyfills if it is using old ng style polyfills', async () => {
+    tree.write('apps/app1/src/polyfills.ts', 'import zone.js;');
+    const project = readProjectConfiguration(tree, 'app1');
+    project.targets.build.options.polyfills = 'apps/app1/src/polyfills.ts';
+    updateProjectConfiguration(tree, 'app1', project);
+
+    expect(tree.exists('karma.conf.js')).toBeFalsy();
+    await karmaProjectGenerator(tree, { project: 'app1' });
+
+    expect(tree.exists('/apps/app1/tsconfig.spec.json')).toBeTruthy();
+    expect(
+      tree.read('/apps/app1/tsconfig.spec.json', 'utf-8')
+    ).toMatchSnapshot();
   });
 
   it('should create a karma.conf.js', async () => {
@@ -121,7 +153,7 @@ describe('karmaProject', () => {
         builder: '@angular-devkit/build-angular:karma',
         options: {
           main: 'apps/app1/src/test.ts',
-          polyfills: 'apps/app1/src/polyfills.ts',
+          polyfills: ['zone.js', 'zone.js/testing'],
           tsConfig: 'apps/app1/tsconfig.spec.json',
           karmaConfig: 'apps/app1/karma.conf.js',
           styles: [],
@@ -132,6 +164,49 @@ describe('karmaProject', () => {
     });
 
     it('should create a tsconfig.spec.json', async () => {
+      await karmaProjectGenerator(tree, { project: 'app1' });
+
+      const tsConfig = devkit.readJson(tree, 'apps/app1/tsconfig.spec.json');
+      expect(tsConfig).toEqual({
+        extends: './tsconfig.json',
+        compilerOptions: {
+          outDir: '../../dist/out-tsc',
+          types: ['jasmine', 'node'],
+        },
+        files: ['src/test.ts'],
+        include: ['**/*.spec.ts', '**/*.test.ts', '**/*.d.ts'],
+      });
+    });
+
+    it('should update the workspace config correctly when using old style ng polyfills', async () => {
+      tree.write('apps/app1/src/polyfills.ts', 'import zone.js;');
+      const project = readProjectConfiguration(tree, 'app1');
+      project.targets.build.options.polyfills = 'apps/app1/src/polyfills.ts';
+      updateProjectConfiguration(tree, 'app1', project);
+
+      await karmaProjectGenerator(tree, { project: 'app1' });
+
+      const workspaceJson = devkit.readJson(tree, 'workspace.json');
+      expect(workspaceJson.projects.app1.architect.test).toEqual({
+        builder: '@angular-devkit/build-angular:karma',
+        options: {
+          main: 'apps/app1/src/test.ts',
+          polyfills: 'apps/app1/src/polyfills.ts',
+          tsConfig: 'apps/app1/tsconfig.spec.json',
+          karmaConfig: 'apps/app1/karma.conf.js',
+          styles: [],
+          scripts: [],
+          assets: [],
+        },
+      });
+    });
+
+    it('should create a tsconfig.spec.json  when using old style ng polyfills', async () => {
+      tree.write('apps/app1/src/polyfills.ts', 'import zone.js;');
+      const project = readProjectConfiguration(tree, 'app1');
+      project.targets.build.options.polyfills = 'apps/app1/src/polyfills.ts';
+      updateProjectConfiguration(tree, 'app1', project);
+
       await karmaProjectGenerator(tree, { project: 'app1' });
 
       const tsConfig = devkit.readJson(tree, 'apps/app1/tsconfig.spec.json');

--- a/packages/angular/src/generators/karma-project/lib/update-tsconfig.ts
+++ b/packages/angular/src/generators/karma-project/lib/update-tsconfig.ts
@@ -8,7 +8,7 @@ import {
 export function updateTsConfigs(tree: Tree, project: string): void {
   const projectConfig = readProjectConfiguration(tree, project);
 
-  return updateJson(
+  updateJson(
     tree,
     joinPathFragments(projectConfig.root, 'tsconfig.json'),
     (json) => {
@@ -20,6 +20,30 @@ export function updateTsConfigs(tree: Tree, project: string): void {
             path: './tsconfig.spec.json',
           },
         ],
+      };
+    }
+  );
+
+  let extraFiles: string[] = [];
+  if (
+    projectConfig.projectType == 'application' &&
+    projectConfig.targets.build?.options?.polyfills &&
+    typeof projectConfig.targets.build.options.polyfills === 'string'
+  ) {
+    let polyfillsPath = projectConfig.targets.build.options.polyfills;
+    polyfillsPath = polyfillsPath.startsWith(projectConfig.root)
+      ? polyfillsPath.replace(`${projectConfig.root}/`, '')
+      : polyfillsPath;
+    extraFiles = [polyfillsPath];
+  }
+
+  return updateJson(
+    tree,
+    joinPathFragments(projectConfig.root, 'tsconfig.spec.json'),
+    (json) => {
+      return {
+        ...json,
+        files: [...json.files, ...extraFiles],
       };
     }
   );

--- a/packages/angular/src/generators/karma-project/lib/update-tsconfig.ts
+++ b/packages/angular/src/generators/karma-project/lib/update-tsconfig.ts
@@ -8,7 +8,7 @@ import {
 export function updateTsConfigs(tree: Tree, project: string): void {
   const projectConfig = readProjectConfiguration(tree, project);
 
-  updateJson(
+  return updateJson(
     tree,
     joinPathFragments(projectConfig.root, 'tsconfig.json'),
     (json) => {
@@ -20,19 +20,6 @@ export function updateTsConfigs(tree: Tree, project: string): void {
             path: './tsconfig.spec.json',
           },
         ],
-      };
-    }
-  );
-
-  const extraFiles =
-    projectConfig.projectType === 'library' ? [] : ['src/polyfills.ts'];
-  return updateJson(
-    tree,
-    joinPathFragments(projectConfig.root, 'tsconfig.spec.json'),
-    (json) => {
-      return {
-        ...json,
-        files: [...json.files, ...extraFiles],
       };
     }
   );

--- a/packages/angular/src/generators/karma-project/lib/update-workspace-config.ts
+++ b/packages/angular/src/generators/karma-project/lib/update-workspace-config.ts
@@ -19,7 +19,6 @@ export function updateWorkspaceConfig(tree: Tree, project: string): void {
   if (projectConfig.projectType === 'application') {
     projectConfig.targets.test.options = {
       ...projectConfig.targets.test.options,
-      polyfills: joinPathFragments(projectConfig.sourceRoot, 'polyfills.ts'),
       styles: [],
       scripts: [],
       assets: [],

--- a/packages/angular/src/generators/karma-project/lib/update-workspace-config.ts
+++ b/packages/angular/src/generators/karma-project/lib/update-workspace-config.ts
@@ -17,8 +17,13 @@ export function updateWorkspaceConfig(tree: Tree, project: string): void {
   };
 
   if (projectConfig.projectType === 'application') {
+    const polyfills = projectConfig.targets.build?.options?.polyfills;
+    let polyfillsPath =
+      polyfills && typeof polyfills === 'string' ? polyfills : undefined;
+
     projectConfig.targets.test.options = {
       ...projectConfig.targets.test.options,
+      polyfills: polyfillsPath ?? ['zone.js', 'zone.js/testing'],
       styles: [],
       scripts: [],
       assets: [],

